### PR TITLE
Begin implementation of TopHitsAggregation

### PR DESCRIPTION
**Pull request by Sean-Der**
As mentioned in issue #64 elastic.js currently doesn't have TopHitsAggregation

This doesn't have any tests, but I will add them in the next day or two. I would like to hear if this is the follow proper patterns, and if not what I should fix up before I write tests that depend on this code.

thanks!

ps: @babeya would you mind checking this out and see if it is as you expect?

### CODE REVIEW

File: src/aggregations/TopHitsAggregation.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    <p>top_hits metric aggregator keeps track of the most relevant document being
+    aggregated.</p>
+
+    @param {String} name The name which be used to refer to this aggregation.
+
+    */
+  ejs.TopHitsAggregation = function (name) {
+
+    var
+      _common = ejs.AggregationMixin(name),
```
Comment (mattweber) [inline src/aggregations/TopHitsAggregation.js]: Top hits in a metrics aggregation you can use the following:

```
_common = ejs.MetricsAggregationMixin(name, 'top_hits')
```

and get rid of

```
agg[name].top_hits = {};
```

**Comment by mattweber** [comment]
This looks great.  I did have one comment and I would like to add the remaining options and tests before pulling this in.  If you have time to do this you can find all the options here:

https://github.com/elasticsearch/elasticsearch/blob/master/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsParser.java

Thanks!

**Comment by Sean-Der** [comment]
Hi @mattweber 

Thanks for checking this out! I will fix the mixin, add tests and add the remaining options in the next couple of days.

Thanks for looking at it, and elastic.js is awesome really enjoy using it.

**Comment by ErwanPigneul** [comment]
Hello, 
I just test it and it's working !
We use this nice library for this project : https://github.com/pagesjaunes/curiosity

Thanks

**Comment by ErwanPigneul** [comment]
This aggregation is particular since it support per hit feature.
I will perhaps try to enrich the agg.

**Comment by Sean-Der** [comment]
@ErwanPigneul 

Hey! Sorry I didn't get to this earlier, I am going to be working on this tonight. Hopefully I will be able to get it all knocked out and and ready for merging tonight.

Have you made any progress? Don't want to duplicate work if not needed.

thanks!

**Comment by ErwanPigneul** [comment]
Hi Sean

no not yet :)

**Comment by Sean-Der** [comment]
@ErwanPigneul 
This should be good to go! If you want to use this in your code base now, the API shouldn't change

@mattweber 
The last handful of functions _should_ be in a mixin, but because of the differing struct of query/aggregations I didn't see a way to do it that follows an existing pattern in the code base (didn't look super close pretty late) Would you mind looking it over, and if I need to move stuff into a mixin just tell me.

Also, this should have full test coverage for the functions defined by TopHitsAggregation

thanks!

**Comment by ErwanPigneul** [comment]
Awesome, thanks a lot @Sean-Der !

**Comment by Sean-Der** [comment]
Hey @mattweber anything I can do to improve this?

thanks

**Comment by mattweber** [comment]
Thanks @Sean-Der!  I have not had a chance to review this yet, hopefully I will be able to get to it this week.  I will keep you updated.  Thanks!

**Comment by mattweber** [comment]
PR looks good.  I am going to pull it into a local branch run some tests then commit in the next day or so.  Thanks again!

**Comment by arpu** [comment]
hello
any idea how i can implemt this with elastic.js ? 

curl -XGET "http://localhost:9200/myjdbc/mytype/_search?search_type=count" -d'
{
  "query": {
    "match": {
       "_all" : "house"
    }
  },
  "aggs": {
    "title": {
      "terms": {
        "field": "title.raw"
      },
      "aggs": {
        "top_tags_hits": {
          "top_hits": {
          }
        },
        "top_hit": {
          "max": {
             "lang" : "expression",
             "script": "_score"
          }
        }
      }
    }
  }
}'

**Comment by Sean-Der** [comment]
Hi @arpu 

This is not meant to be a place to ask for support, it makes it difficult for others when you go off topic. However, I have implemented your query below.

``` javascript
var request = ejs.Request().query(ejs.MatchQuery('_all', 'house'));                                                                                                                                                                            
request = request.agg(ejs.TermsAggregation('title').                                                                                                                                                                                           
                        field('title.raw').                                                                                                                                                                                                    
                        agg(ejs.TopHitsAggregation('top_tags_hits')).                                                                                                                                                                          
                        agg(ejs.MaxAggregation('top_hit').lang('expression').script('_score')));  
```

``` json
{  
   "query":{  
      "match":{  
         "_all":{  
            "query":"house"
         }
      }
   },
   "aggs":{  
      "title":{  
         "terms":{  
            "field":"title.raw"
         },
         "aggs":{  
            "top_tags_hits":{  },
            "top_hit":  {
               "max":{  
                  "lang":"expression",
                  "script":"_score"
               }
         }
      }
   }
}
```

**Comment by arpu** [comment]
hello @Sean-Der 

sorry for asking this in the pull request, where is the recommended place to ask user questions? 

thx a lot for implemend the query! 

best regards

**Comment by mattweber** [comment]
Looks great, merged!  Thanks!

**Pull request merged**